test/versioned/fastify/add-hook.tap.js

```diff
@@ -0,0 +1,144 @@
+  })
+}
+
+tap.test('fastify hook instrumentation', (t) => {
+  t.autoend()
+
+  let agent = null
+  let fastify = null
+
+  t.beforeEach(() => {
+    agent = helper.instrumentMockedAgent({
+      feature_flag: {
+        fastify_instrumentation: true
+      }
+    })
+    fastify = require('fastify')()
+  })
+
+  t.afterEach(() => {
+    helper.unloadAgent(agent)
+    fastify.close()
+  })
+
+  t.test('non-error hooks', async function nonErrorHookTest(t) {
+    // setup fastify route
+    fastify.get('/', function routeHandler(_, reply) {
+      t.comment('route handler called')
+      reply.send({ hello: 'world' })
+    })
+
+    // setup hooks
+    const ok = ALL_HOOKS.reduce((all, hookName) => {
+      t.comment(`registering ${hookName}`)
+      all[hookName] = false
+      return all
+    }, {})
+
+    ALL_HOOKS.forEach((hookName) => {
+      fastify.addHook(hookName, function testHook(...args) {
+        t.comment(`${hookName} called`)
+        // lifecycle signatures vary between the events
+        // the last arg is always the next function though
+        const next = args[args.length - 1]
+        ok[hookName] = true
+        next()
+      })
+    })
+
+    agent.on('transactionFinished', (transaction) => {
+      t.equal('WebFrameworkUri/Fastify/GET//', transaction.getName(), `transaction name matched`)
+      // all the hooks are siblings of the route handler
+      // except the AFTER_HANDLER_HOOKS which are children of the route handler
+      metrics.assertSegments(transaction.trace.root, [
+        'WebTransaction/WebFrameworkUri/Fastify/GET//',
+        [
+          ...getSegmentNames(REQUEST_HOOKS),
+          'Nodejs/Middleware/Fastify/routeHandler',
+          getSegmentNames(AFTER_HANDLER_HOOKS)
+        ]
+      ])
+    })
+
+    await fastify.listen(0)
+    const { port } = fastify.server.address()
+    const result = await getAsync(`http://127.0.0.1:${port}/`)
+    t.equal(result.statusCode, 200)
+    t.equal(result.body, JSON.stringify({ hello: 'world' }))
+
+    // verify every hook was called after response
+    for (const [hookName, isOk] of Object.entries(ok)) {
+      t.equal(isOk, true, `${hookName} captured`)
+    }
+    t.end()
+  })
+
+  t.test('error hook', async function errorHookTest(t) {
+    // setup fastify route
+    fastify.get('/', function routeHandler() {
+      throw new Error('sup')
+    })
+
+    const hookName = 'onError'
+    let ok = false
+
+    fastify.addHook(hookName, (req, reply, next) => {
+      ok = true
+      next()
+    })
+
+    agent.on('transactionFinished', (transaction) => {
+      t.equal('WebFrameworkUri/Fastify/GET//', transaction.getName(), `transaction name matched`)
+      // all the hooks are siblings of the route handler
+      metrics.assertSegments(transaction.trace.root, [
+        'WebTransaction/WebFrameworkUri/Fastify/GET//',
+        [
+          'Nodejs/Middleware/Fastify/routeHandler'
+          // `Nodejs/Middleware/Fastify/${hookName}/testHook`
```

Comment:
I would expect the onError hook to be listed as a segment but... it isn't? Should I be worried?

Reply:
it should def be there but the reason it is not is because in the instrumentation you didnt' add `onError` into the valid `REQUEST_HOOKS`. I added and i see the segment

<img width="1202" alt="screenshot 2021-10-07 at 10 42 18 AM" src="https://user-images.githubusercontent.com/1874937/136407677-1e6bf2fc-4d83-4665-b3d0-41d7d2617fce.png">